

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Overview &mdash; MineSV 0.4 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Run MineSV demo at Biowulf" href="cookbooks/MineSV_demo_at_Biowulf.html" />
    <link rel="prev" title="Welcome to MineSV’s documentation!" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> MineSV
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#i-description">I.  Description</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ii-a-little-history-about-minesv">II. A little history about MineSV</a></li>
<li class="toctree-l2"><a class="reference internal" href="#iii-dependencies">III.  Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="#iv-analysis-modes">IV.  Analysis modes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v-run-modes">V.  Run modes</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vi-callers-available">VI.  Callers available</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#user-s-guide">User’s guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#i-input-requirements">I.  Input requirements</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#a-required-for-all-run-modes">A.  Required for all run modes:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#b-required-for-callonly-or-callandannotate-run-modes">B.  Required for <strong>callOnly</strong> or <strong>callAndAnnotate</strong> run modes:</a></li>
<li class="toctree-l3"><a class="reference internal" href="#c-required-for-annotateonly-run-mode">C.  Required for annotateOnly run mode:</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ii-sample-file-formats">II.  Sample file formats</a></li>
<li class="toctree-l2"><a class="reference internal" href="#iii-comparison-and-annotation">III.  Comparison and annotation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#a-sv-comparison-methods">A. SV comparison methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="#b-annotation">B. Annotation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#iv-editing-the-config-yaml">IV.  Editing the config.yaml</a></li>
<li class="toctree-l2"><a class="reference internal" href="#v-to-run">V.  To run</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vi-to-monitor-pipeline">VI.  To monitor pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vi-output">VI. Output</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#developer-s-guide">Developer’s guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#i-pipeline-architecture">I.  Pipeline architecture</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ii-to-add-callers">II.  To add callers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#citations">Citations</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Cookbook</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="cookbooks/MineSV_demo_at_Biowulf.html">Run MineSV demo at Biowulf</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MineSV</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Overview</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/readme.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<section id="i-description">
<h2>I.  Description<a class="headerlink" href="#i-description" title="Permalink to this headline">¶</a></h2>
<p>This pipeline coordinates and monitors the submission of a list of bams for structural variant discovery, and comparison/annotation of results.  You may opt to run calling alone, annotation alone, or both sequentially.  Multiple callers are included with MineSV, and additional callers may be added.  The pipeline may be run on an HPC or in a local environment.  Details are described in subsequent sections, and the main documentation is available at <a class="reference external" href="https://nci-cgr.github.io/MineSV/">https://nci-cgr.github.io/MineSV/</a> .</p>
</section>
<section id="ii-a-little-history-about-minesv">
<h2>II. A little history about MineSV<a class="headerlink" href="#ii-a-little-history-about-minesv" title="Permalink to this headline">¶</a></h2>
<p>MineSV was started to be developed at the NCI-Cancer Genomics Research Laboratory (NCI-CGR) in 2018, with the original name <a href="#id1"><span class="problematic" id="id2">**</span></a><em>MoCCA-SV</em>** (Modular Calling, Comparison, and Annotation of Structural Variants). It has been renamed as <a href="#id3"><span class="problematic" id="id4">**</span></a><em>MineSV</em>** since June 2021.</p>
</section>
<section id="iii-dependencies">
<h2>III.  Dependencies<a class="headerlink" href="#iii-dependencies" title="Permalink to this headline">¶</a></h2>
<p>To run MineSV, you will need:</p>
<ul class="simple">
<li><p>python v3</p></li>
<li><p>snakemake</p></li>
<li><p>perl v5.18+</p></li>
<li><p>singularity v2+</p></li>
</ul>
</section>
<section id="iv-analysis-modes">
<h2>IV.  Analysis modes<a class="headerlink" href="#iv-analysis-modes" title="Permalink to this headline">¶</a></h2>
<p>One of the following analysis modes may be specified in the configuration file:</p>
<ol class="arabic simple">
<li><p>Tumor/Normal: for matched tumor/normal pairs; set analysisMode to ‘TN’</p></li>
<li><p>Tumor only: for tumor samples with no paired normal; set analysisMode to ‘TO’ (under development)</p></li>
<li><p>De novo: for trios to detect de novo SVs in the proband; set analysisMode to ‘de_novo’</p></li>
<li><p>Germline: for germline samples; set analysisMode to ‘germline’</p></li>
</ol>
</section>
<section id="v-run-modes">
<h2>V.  Run modes<a class="headerlink" href="#v-run-modes" title="Permalink to this headline">¶</a></h2>
<p>The pipeline has three run modes available; select one in the configuration file under “runMode”:</p>
<ol class="arabic simple">
<li><p>callOnly: perform SV calling on user-specified bams</p></li>
<li><p>annotateOnly: annotate and compare user-specified SV caller outputs</p></li>
<li><p>callAndAnnotate: call SVs and annotate (user only has to specify the initial bam inputs for this mode; caller outputs are automatically detected)</p></li>
</ol>
</section>
<section id="vi-callers-available">
<h2>VI.  Callers available<a class="headerlink" href="#vi-callers-available" title="Permalink to this headline">¶</a></h2>
<p>Select from the following callers in the configuration file under “callers”:</p>
<ul class="simple">
<li><p>svaba</p></li>
<li><p>manta</p></li>
<li><p>delly</p></li>
<li><p>breakdancer</p></li>
</ul>
<p>To add a caller, three new elements are required: a snakefile to run the calling, a caller_to_bed script to convert the output to bed format, and a container recipe file with the caller installed. Please see the developer’s guide below for more details.</p>
</section>
</section>
<section id="user-s-guide">
<h1>User’s guide<a class="headerlink" href="#user-s-guide" title="Permalink to this headline">¶</a></h1>
<section id="i-input-requirements">
<h2>I.  Input requirements<a class="headerlink" href="#i-input-requirements" title="Permalink to this headline">¶</a></h2>
<section id="a-required-for-all-run-modes">
<h3>A.  Required for all run modes:<a class="headerlink" href="#a-required-for-all-run-modes" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Edited config.yaml (note that you can customize the name and location of this file)</p></li>
<li><p>Sample file (required format and contents depend on run mode and analysis mode; see subsequent section for details)</p></li>
</ul>
</section>
<section id="b-required-for-callonly-or-callandannotate-run-modes">
<h3>B.  Required for <strong>callOnly</strong> or <strong>callAndAnnotate</strong> run modes:<a class="headerlink" href="#b-required-for-callonly-or-callandannotate-run-modes" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Reference genome (with our without indices; the pipeline will generate any missing index files)</p></li>
<li><p>Sorted and indexed bam files</p></li>
</ul>
</section>
<section id="c-required-for-annotateonly-run-mode">
<h3>C.  Required for annotateOnly run mode:<a class="headerlink" href="#c-required-for-annotateonly-run-mode" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Caller output</p></li>
</ul>
</section>
</section>
<section id="ii-sample-file-formats">
<h2>II.  Sample file formats<a class="headerlink" href="#ii-sample-file-formats" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p>Sample files for <strong>callOnly</strong> or <strong>callAndAnnotate</strong> run modes:</p>
<ul>
<li><p>Tumor/normal mode:</p>
<ul class="simple">
<li><p>Three space-delimited, headerless columns: sample name, tumor bam, normal bam</p></li>
<li><p>File names only - not path (the path is specified in config.yaml as “inDir”)</p>
<ul>
<li><p>The pipeline is currently set up such that all bams are assumed to be in the same directory</p></li>
</ul>
</li>
<li><p>Example:</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="mi">301</span><span class="n">_pair</span> <span class="mi">301_4</span><span class="n">E_tumor</span><span class="o">.</span><span class="n">bam</span> <span class="mi">301_2</span><span class="n">A_blood</span><span class="o">.</span><span class="n">bam</span>
<span class="mi">275</span><span class="n">_pair</span> <span class="mi">275_1</span><span class="n">B_tumor</span><span class="o">.</span><span class="n">bam</span> <span class="mi">275_3</span><span class="n">A_blood</span><span class="o">.</span><span class="n">bam</span>
<span class="mi">275</span><span class="n">met_pair</span> <span class="mi">275_2</span><span class="n">C_tumor</span><span class="o">.</span><span class="n">bam</span> <span class="mi">275_3</span><span class="n">A_blood</span><span class="o">.</span><span class="n">bam</span>
<span class="n">Best_pair</span> <span class="mi">123_3</span><span class="n">S</span><span class="o">.</span><span class="n">bam</span> <span class="mi">456_2</span><span class="n">M</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
</li>
<li><p>De novo mode:</p>
<ul class="simple">
<li><p>Four space-delimited, headerless columns: family ID, parent1 bam, parent2 bam, child bam</p></li>
<li><p>File names only</p></li>
<li><p>Example:</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Family_A</span> <span class="n">mom</span><span class="o">.</span><span class="n">bam</span> <span class="n">dad</span><span class="o">.</span><span class="n">bam</span> <span class="n">child</span><span class="o">.</span><span class="n">bam</span>
<span class="n">F1234_1</span> <span class="n">parentA</span><span class="o">.</span><span class="n">bam</span> <span class="n">parentB</span><span class="o">.</span><span class="n">bam</span> <span class="n">child_1</span><span class="o">.</span><span class="n">bam</span>
<span class="n">F1234_2</span> <span class="n">parentA</span><span class="o">.</span><span class="n">bam</span> <span class="n">parentB</span><span class="o">.</span><span class="n">bam</span> <span class="n">child_2</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
</li>
<li><p>Tumor only mode:</p></li>
<li><p>Germline mode:</p>
<ul class="simple">
<li><p>Two space-delimited, headerless columns: subject ID, bam</p></li>
<li><p>File name only (no path)</p></li>
<li><p>Example:</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Patient1</span> <span class="mf">15324.</span><span class="n">bam</span>
<span class="n">Patient2</span> <span class="n">pt2</span><span class="o">.</span><span class="n">bam</span>
</pre></div>
</div>
</li>
</ul>
</li>
<li><p>Sample files for <strong>annotateOnly</strong> mode:</p>
<ul class="simple">
<li><p>Note that this sample file is automatically generated if MineSV is used to call the data</p></li>
<li><p>Space-delimited with header line</p></li>
<li><p>Header: sample caller1 caller2 …</p></li>
<li><p>File names with full paths</p></li>
<li><p>Example:</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span> <span class="n">svaba</span> <span class="n">delly</span>
<span class="n">A101</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SV_out1</span><span class="o">.</span><span class="n">vcf</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">A101</span><span class="o">.</span><span class="n">out</span>
</pre></div>
</div>
</li>
</ul>
</section>
<section id="iii-comparison-and-annotation">
<h2>III.  Comparison and annotation<a class="headerlink" href="#iii-comparison-and-annotation" title="Permalink to this headline">¶</a></h2>
<section id="a-sv-comparison-methods">
<h3>A. SV comparison methods<a class="headerlink" href="#a-sv-comparison-methods" title="Permalink to this headline">¶</a></h3>
<p>We employ two different methods to compare SVs across callers and to public SV databases.  For intra-chromosomal SVs, we use percent reciprocal overlap (e.g., to be considered comparable, SV1 must overlap at least X% of SV2, and SV2 must overlap at least X% of SV1).  For intra-chromosomal insertions, we first pad by a number of bases up- and down-stream, then apply percent reciprocal overlap.  For inter-chromosomal SVs, we look for overlap within a window surrounding each breakend.</p>
</section>
<section id="b-annotation">
<h3>B. Annotation<a class="headerlink" href="#b-annotation" title="Permalink to this headline">¶</a></h3>
<p>We include three categories of annotation in the final output file:</p>
<ol class="arabic simple">
<li><p>Genomic context (non-reciprocal percent overlap with regions in UCSC browser’s RepeatMasker, SegDups, and Telomere/Centromere tracks)</p></li>
<li><p>Public data (identity, as defined in section A, with SVs reported by 1000 Genomes, ClinGen, ClinVar, and DGV)</p>
<ul class="simple">
<li><p>Insertions (defined as SVs &lt;10bp) in the public datasets have been padded by +/- 500bp</p></li>
<li><p>Three public datasets are pulled from dbvar</p>
<ul>
<li><p>1000 Genomes (estd219): 66k variant regions from healthy individuals in the 1000 Genomes Consortium Phase3 Integrated SV dataset</p></li>
<li><p>ClinVar (nstd102): 765 SVs with clinical assertions</p></li>
<li><p>ClinGen (nstd37): laboratory-submitted data - 33,378 SVs with clinical assertions as classified by the original submitter</p></li>
</ul>
</li>
<li><p>DGV data is from the relevant UCSC browser track and consists of intra-chromosomal SVs from healthy individuals</p></li>
</ul>
</li>
<li><p>Geneset (any amount of overlap with RefSeq transcripts/gene names)</p></li>
</ol>
<p>See our publication (<em>__</em>) for additional details.  Parameters defining these comparisons are user-configurable; see subsequent section.</p>
</section>
</section>
<section id="iv-editing-the-config-yaml">
<h2>IV.  Editing the config.yaml<a class="headerlink" href="#iv-editing-the-config-yaml" title="Permalink to this headline">¶</a></h2>
<p><strong>Basic analysis parameters</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">analysisMode:</span></code> Enter one value for analysisMode (TN, TO, germline, de_novo)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">callers:</span></code> Select callers to run by providing a list in YAML format</p>
<ul>
<li><p>The name used for the caller in the config file must match (including case) the caller name used in <code class="docutils literal notranslate"><span class="pre">modules/Snakefile_&lt;caller&gt;_&lt;runMode&gt;</span></code> and in <code class="docutils literal notranslate"><span class="pre">scripts/&lt;caller&gt;_to_bed.sh</span></code>, e.g.:<span class="raw-html-m2r"><br></span>
modules/Snakefile_DELLY_TN, scripts/delly_to_bed.sh, config file ‘Delly2’: NO<span class="raw-html-m2r"><br></span>
modules/Snakefile_delly_TN, scripts/delly_to_bed.sh, config file ‘delly’: YES</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">runMode:</span></code> Select one runMode by setting one option to ‘yes’ and the others to ‘no’</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">refGenome:</span></code> Provide the full path and filename for a reference fasta</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">genomeBuild:</span></code> Record the genome build (either ‘hg19’ or ‘b37’)</p></li>
</ul>
<p><strong>Annotation parameters</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">annotateFile:</span></code> If using annotateOnly mode, provide the <code class="docutils literal notranslate"><span class="pre">/full/path/to/SV_files_for_annotation.txt</span></code></p>
<ul>
<li><p>This is the file indicating which caller outputs to compare and annotate.  If MineSV was used to generate the caller outputs, the file will be created for you and named SV_files_for_annotation.txt; otherwise you can manually create and name your own file in this format.</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">annotationParams:</span></code> This section allows you to customize the stringency with which various comparisons are made.</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">interchromPadding:</span></code> base pairs to add on either side of the breakend to be used to find overlap when comparing inter-chromosomal SVs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">insertionPadding:</span></code> base pairs to add on either side of an insertion to find overlap when comparing across callers and annotating</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">crossCallerOverlap:</span></code> minimum reciprocal overlap required to count as an intersection in other callers’ data on the same subject</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">genomicContextOverlap:</span></code> what % of an SV must reside in a queried region (repeat, segdup, centro/telomeric) to count as an intersection</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">publicDataOverlap:</span></code> minimum reciprocal overlap required to count as an intersection (DGV, ClinVar, ClinGen, 1KG)</p></li>
</ul>
</li>
</ul>
<p><strong>Samples to analyze</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">inFile:</span></code> Provide the full path and sample file that is appropriate for the analysis mode (TN, TO, germline, de_novo); N/A if using annotateOnly mode</p></li>
</ul>
<p><strong>Directories</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">inDir:</span></code> Location of the input BAM files (N/A if annotateOnly)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">execDir:</span></code> Location of pipeline</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">outDir:</span></code> Desired location for output files</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">logDir:</span></code> Desired location for log files</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tempDir:</span></code> Temporary or scratch directory, used for rapidly created and deleted files generated by parallelizable callers</p></li>
</ul>
<p><strong>Cluster parameters</strong></p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">clusterMode:</span></code></p>
<ul>
<li><p>If you are running on an HPC, provide the string you want to use to submit jobs.  Examples:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;qsub -V -S /bin/sh -q queue.q -o /my/path/SV_out/logs -j y&#39;</span>
<span class="c1"># or</span>
<span class="s1">&#39;sbatch --get-user-env -D /my/path/SV_out --mem=8g&#39;</span>
</pre></div>
</div>
</li>
<li><p>If you are not on a cluster, you can run in local mode by entering <code class="docutils literal notranslate"><span class="pre">'local'</span></code></p></li>
<li><p>If snakemake has locked the working directory, you can unlock by entering <code class="docutils literal notranslate"><span class="pre">'unlock'</span></code></p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">latency:</span></code> Enter the number of seconds to wait for snakemake to detect rule output.  Use 300+ if running on a cluster.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">maxNumJobs:</span></code> Enter the maximum number of jobs to run in parallel; N/A if running locally</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">maxThreads:</span></code> Assign the number of threads for each caller listed in <code class="docutils literal notranslate"><span class="pre">callers:</span></code> above, in matching order.  For callers that can not use multi-threading, assign 1.  Assign 1 if running locally and/or without multi-threading.</p></li>
</ul>
</section>
<section id="v-to-run">
<h2>V.  To run<a class="headerlink" href="#v-to-run" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Copy config.yaml to your directory</p></li>
<li><p>Edit config.yaml and save</p></li>
<li><p>Run <code class="docutils literal notranslate"><span class="pre">SV_wrapper.sh</span> <span class="pre">/full/path/config.yaml</span></code> as appropriate for your environment (e.g. on an HPC using the SGE job scheduler, run <code class="docutils literal notranslate"><span class="pre">qsub</span> <span class="pre">-q</span> <span class="pre">&lt;queue&gt;</span> <span class="pre">SV_wrapper.sh</span> <span class="pre">/full/path/config.yaml</span></code>)</p></li>
</ul>
</section>
<section id="vi-to-monitor-pipeline">
<h2>VI.  To monitor pipeline<a class="headerlink" href="#vi-to-monitor-pipeline" title="Permalink to this headline">¶</a></h2>
<ul>
<li><p>Look in log directory (specified in config file) for snakemake logged stdout <code class="docutils literal notranslate"><span class="pre">MineSV_DATETIME.out</span></code></p></li>
<li><p>Look in same log directory as above for logs for each rule (e.g. <code class="docutils literal notranslate"><span class="pre">snakejob.delly_call.119.sh.o123456</span></code>; file names will vary depeneding on environment)</p></li>
<li><p>Look in log directory for SV_wrapper.log.DATETIME.  Example output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">config</span><span class="o">.</span><span class="n">yaml</span> <span class="n">passes</span> <span class="nb">all</span> <span class="n">checks</span><span class="o">.</span>
<span class="n">Command</span> <span class="n">run</span><span class="p">:</span> <span class="n">conf</span><span class="o">=/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">config</span><span class="o">.</span><span class="n">yaml</span> <span class="n">snakemake</span> <span class="o">-</span><span class="n">p</span> <span class="o">-</span><span class="n">s</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">Snakefile_SV_scaffold</span> <span class="o">--</span><span class="n">use</span><span class="o">-</span><span class="n">singularity</span> <span class="o">--</span><span class="n">singularity</span><span class="o">-</span><span class="n">args</span> <span class="s2">&quot;-B /path/to/input:/input,/ttemp/sv:/scratch,/path/to/refGenomes:/ref,/path/to/SV_out:/output,/path/to/pipeline:/exec&quot;</span> <span class="o">--</span><span class="n">rerun</span><span class="o">-</span><span class="n">incomplete</span> <span class="o">--</span><span class="n">cluster</span> <span class="s2">&quot;qsub -V -S /bin/sh -q queue.q -o /path/to/SV_out/logs -j y&quot;</span> <span class="o">--</span><span class="n">jobs</span> <span class="mi">8</span> <span class="o">--</span><span class="n">latency</span><span class="o">-</span><span class="n">wait</span> <span class="mi">300</span> <span class="o">&amp;&gt;</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">SV_out</span><span class="o">/</span><span class="n">logs</span><span class="o">/</span><span class="n">MineSV_201904041317</span><span class="o">.</span><span class="n">out</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Note that for troubleshooting, you can re-run the command printed in this log with additional snakemake command line options, like <code class="docutils literal notranslate"><span class="pre">-n</span></code> for a dry run or <code class="docutils literal notranslate"><span class="pre">--dag</span> <span class="pre">...</span> <span class="pre">|</span> <span class="pre">dot</span> <span class="pre">-Tsvg</span> <span class="pre">&gt;</span> <span class="pre">dag.svg</span></code> to visualize the directed acyclic graph.  See snakemake documentation for more details.</p></li>
</ul>
</li>
</ul>
</section>
<section id="vi-output">
<h2>VI. Output<a class="headerlink" href="#vi-output" title="Permalink to this headline">¶</a></h2>
<p>MineSV has two main outputs: SV calls from each caller, and the union of all callers for each sample along with comparison and annotation information.</p>
<ol class="arabic simple">
<li><p>Each caller run will write to its own directory (e.g. delly run in tumor/normal mode will create and write to a directory called <code class="docutils literal notranslate"><span class="pre">delly_TN</span></code>).  See the individual callers’ documentation for descriptions of output.</p></li>
<li><p>Running in callAndAnnotate or callOnly modes will create the directory <code class="docutils literal notranslate"><span class="pre">compare_and_annotate</span></code>.  This directory contains the files <code class="docutils literal notranslate"><span class="pre">intrachromosomal_SVs_&lt;sample&gt;</span></code>, which contain the superset of SVs from all callers.  <em>NOTE: this is under development, and inter-chromosomal SVs are being integrated currently.</em></p></li>
</ol>
<p>Example output directory structure is as follows.  Exact output will depend upon run mode and analysis mode selected.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">|---</span> <span class="n">user</span><span class="o">/</span><span class="n">defined</span><span class="o">/</span><span class="n">output_dir</span><span class="o">/</span>
   <span class="o">|---</span> <span class="n">delly_TN</span><span class="o">/</span>
   <span class="o">|---</span> <span class="n">svaba_TN</span><span class="o">/</span>
   <span class="o">|---</span> <span class="n">logs</span><span class="o">/</span>
   <span class="o">|</span>  <span class="o">|---</span> <span class="n">MineSV_</span><span class="o">&lt;</span><span class="n">datetime</span><span class="o">&gt;.</span><span class="n">out</span>
   <span class="o">|---</span> <span class="n">SV_files_for_annotation</span><span class="o">.</span><span class="n">txt</span>
   <span class="o">|---</span> <span class="n">compare_and_annotate</span><span class="o">/</span>
      <span class="o">|---</span> <span class="n">intrachromosomal_SVs_sampleA</span>
      <span class="o">|---</span> <span class="n">intrachromosomal_SVs_sampleB</span>
</pre></div>
</div>
<p>Headers from file <code class="docutils literal notranslate"><span class="pre">intrachromosomal_SVs_&lt;sample&gt;</span></code>:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Field</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>#CHROM</p></td>
<td><p>chromosome</p></td>
</tr>
<tr class="row-odd"><td><p>pos1</p></td>
<td><p>upstream breakend</p></td>
</tr>
<tr class="row-even"><td><p>pos2</p></td>
<td><p>downstream breakend</p></td>
</tr>
<tr class="row-odd"><td><p>svaba</p></td>
<td><p>count of comparable SVs in svaba (based on % reciprocal overlap); “orig” if this SV originated from this caller</p></td>
</tr>
<tr class="row-even"><td><p>breakdancer</p></td>
<td><p>count of comparable SVs in svaba (based on % reciprocal overlap); “orig” if this SV originated from this caller</p></td>
</tr>
<tr class="row-odd"><td><p>delly</p></td>
<td><p>count of comparable SVs in svaba (based on % reciprocal overlap); “orig” if this SV originated from this caller</p></td>
</tr>
<tr class="row-even"><td><p>manta</p></td>
<td><p>count of comparable SVs in svaba (based on % reciprocal overlap); “orig” if this SV originated from this caller</p></td>
</tr>
<tr class="row-odd"><td><p>caller_count</p></td>
<td><p>total number of callers reporting a comparable SV at this position</p></td>
</tr>
<tr class="row-even"><td><p>match_region</p></td>
<td><p>chrom and outer coordinates of a matching pair of SVs in the format chr:end1-end2; serves as a key to find pairwise matches in the superset call data</p></td>
</tr>
<tr class="row-odd"><td><p>RepeatMasker</p></td>
<td><p>count of overlaps between SV and repeat regions</p></td>
</tr>
<tr class="row-even"><td><p>SegDups</p></td>
<td><p>count of overlaps between SV and segmentally duplicated regions</p></td>
</tr>
<tr class="row-odd"><td><p>Telo_Centro</p></td>
<td><p>count of overlaps between SV and telomeric or centromeric regions</p></td>
</tr>
<tr class="row-even"><td><p>1KG_sv_type</p></td>
<td><p>SV type reported for each overlapping SV (e.g. “copy_number_gain”)</p></td>
</tr>
<tr class="row-odd"><td><p>1KG_pheno</p></td>
<td><p>phenotype reported for each overlapping SV (“healthy” for 1000 Genomes)</p></td>
</tr>
<tr class="row-even"><td><p>1KG_clinical_assertion</p></td>
<td><p>determination of clinical significance of SV (“na” for 1000 Genomes)</p></td>
</tr>
<tr class="row-odd"><td><p>ClinGen_sv_type</p></td>
<td><p>SV type reported for each overlapping SV (e.g. “copy_number_gain”)</p></td>
</tr>
<tr class="row-even"><td><p>ClinGen_pheno</p></td>
<td><p>phenotype reported for each overlapping SV (e.g. “Developmental_delay”)</p></td>
</tr>
<tr class="row-odd"><td><p>ClinGen_clinical_assertion</p></td>
<td><p>determination of clinical significance of SV (e.g. “Benign” or “Pathogenic”)</p></td>
</tr>
<tr class="row-even"><td><p>ClinVar_sv_type</p></td>
<td><p>SV type reported for each overlapping SV (e.g. “copy_number_gain”)</p></td>
</tr>
<tr class="row-odd"><td><p>ClinVar_pheno</p></td>
<td><p>phenotype reported for each overlapping SV (e.g. “Developmental_delay”)</p></td>
</tr>
<tr class="row-even"><td><p>ClinVar_clinical_assertion</p></td>
<td><p>determination of clinical significance of SV (e.g. “Benign” or “Pathogenic”)</p></td>
</tr>
<tr class="row-odd"><td><p>DGV_sv_type</p></td>
<td><p>SV type reported for each overlapping SV (e.g. “copy_number_gain”)</p></td>
</tr>
<tr class="row-even"><td><p>DGV_pheno</p></td>
<td><p>phenotype reported for each overlapping SV (“healthy” for DGV)</p></td>
</tr>
<tr class="row-odd"><td><p>DGV_clinical_assertion</p></td>
<td><p>determination of clinical significance of SV (“na” for DGV)</p></td>
</tr>
<tr class="row-even"><td><p>RefSeq_transcripts</p></td>
<td><p>semicolon-delimited list of transcripts overlapping the SV</p></td>
</tr>
<tr class="row-odd"><td><p>RefSeq_genes</p></td>
<td><p>semicolon-delimited list of gene names overlapping SV</p></td>
</tr>
<tr class="row-even"><td><p>Original_caller_output</p></td>
<td><p>all fields from the original caller’s output, delimited by a double underscore “__”</p></td>
</tr>
</tbody>
</table>
<p><em>Think about differences for inter-chrom SVs, e.g. instead of all overlapping genes, just report those that the break ends reside in?  or all genes within x bp?</em></p>
</section>
</section>
<section id="developer-s-guide">
<h1>Developer’s guide<a class="headerlink" href="#developer-s-guide" title="Permalink to this headline">¶</a></h1>
<section id="i-pipeline-architecture">
<h2>I.  Pipeline architecture<a class="headerlink" href="#i-pipeline-architecture" title="Permalink to this headline">¶</a></h2>
<p>MineSV is a Snakemake pipeline controlled by a user-configurable YAML file.  The file Snakefile_SV_scaffold pulls in other sets of rules based on the chosen configuration options, and for either of the two calling run modes, runs a few rules that are required regardless of caller choice.  The scaffold dictates which sets of rules to include based on the analysis mode and the list of callers.  Snakefiles in the <code class="docutils literal notranslate"><span class="pre">modules/</span></code> directory contain the sets of rules that pertain to a given caller and analysis mode.  They are named accordingly, e.g. <code class="docutils literal notranslate"><span class="pre">modules/Snakefile_delly_TN</span></code> contains the rules to run delly on tumor/normal samples.  The scaffold matches the callers and analysis mode listed in the config to the modules/Snakefiles in order to include the applicable rules, but in such a way as to minimize code changes required to add new callers.</p>
</section>
<section id="ii-to-add-callers">
<h2>II.  To add callers<a class="headerlink" href="#ii-to-add-callers" title="Permalink to this headline">¶</a></h2>
<p>There are three steps required to add a caller:</p>
<ol class="arabic simple">
<li><p>Construct a Snakefile, named <code class="docutils literal notranslate"><span class="pre">modules/Snakefile_&lt;caller&gt;_&lt;anMode&gt;</span></code>, to run the new caller in a given analysis mode.  You may start with the appropriate <code class="docutils literal notranslate"><span class="pre">TEMPLATE_Snakefile_caller_&lt;anMode&gt;</span></code> file found in the <code class="docutils literal notranslate"><span class="pre">modules/</span></code> directory.  Note that many variables are sourced from the config file in Snakefile_SV_scaffold and are available in snakefiles that are included from the scaffold.  Be sure that the caller name used in this filename is identical to the caller name you use in the config file.</p></li>
<li><p>Add a shell script to the <code class="docutils literal notranslate"><span class="pre">scripts/</span></code> directory to convert the new caller’s output to bed format.  This can generally be done in three awk statements; to start, see <code class="docutils literal notranslate"><span class="pre">scripts/TEMPLATE_caller_to_bed.sh</span></code> file.  Like in step 1, ensure that you use the same caller name to name this script.</p></li>
<li><p>Create a container with the caller (and any other dependencies) installed.  This can be a singularity container or a docker container, preferably hosted on a public-facing hub.  You may start with <code class="docutils literal notranslate"><span class="pre">TEMPLATE_singularity_recipe</span></code>.  Be sure you are running the relevant rules in your snakefile within the container.</p></li>
</ol>
</section>
</section>
<section id="citations">
<h1>Citations<a class="headerlink" href="#citations" title="Permalink to this headline">¶</a></h1>
<p>The main MineSV documentation is at: <a class="reference external" href="https://nci-cgr.github.io/MineSV/">https://nci-cgr.github.io/MineSV/</a></p>
<p>The original MoCCA-SV pipeline:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/NCI-CGR/MoCCA-SV">https://github.com/NCI-CGR/MoCCA-SV</a></p></li>
</ul>
<p>This pipeline was initially described here:</p>
<ul class="simple">
<li><p>Ballew BJ, Yeager M, Hicks B, Zhu B. MoCCA-SV: A Flexible Ensemble Framework for Structural Variant Analysis. Poster session presented at: 19th Annual General Meeting of Advances in Genome Biology and Technology (AGBT); 2019 Feb 27 – Mar 3, Marco Island FL.</p></li>
</ul>
<p>This pipeline currently uses the following callers:</p>
<ul class="simple">
<li><p>Wala JA et al. SvABA: genome-wide detection of structural variants and indels by local assembly (2018). Genome Res, 28(4):581-91.</p></li>
<li><p>Chen K et al. BreakDancer: an algorithm for high-resolution mapping of genomic structural variation (2009). Nat Methods, 6(9):677-81.</p></li>
<li><p>Rausch T et al. DELLY: structural variant discovery by integrated paired-end and split-read analysis (2012). Bioinformatics, 28(18):i333-i339.</p></li>
<li><p>Chen X et al. Manta: rapid detection of structural variants and indels for germline and cancer sequencing applications (2016). Bioinformatics, 32(8):1220-2.</p></li>
</ul>
<p>This pipeline uses the following workflow management and containerization solutions:</p>
<ul class="simple">
<li><p>Köster J et al. Snakemake–a scalable bioinformatics workflow engine (2012). Bioinformatics, 28(19): 2520–2.</p></li>
<li><p>Kurtzer GM, et al. Singularity: Scientific containers for mobility of compute (2017). PLoS ONE, 12(5): e0177459.</p></li>
<li><p>Sochat V, et al. Enhancing reproducibility in scientific computing: Metrics and registry for Singularity containers (2017). PLoS ONE, 12(11): e0188511.</p></li>
</ul>
</section>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="cookbooks/MineSV_demo_at_Biowulf.html" class="btn btn-neutral float-right" title="Run MineSV demo at Biowulf" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="index.html" class="btn btn-neutral float-left" title="Welcome to MineSV’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Bari Ballew, Wei Zhu.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>