

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Run MineSV demo at Biowulf &mdash; MineSV 0.4 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Overview" href="../readme.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> MineSV
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../readme.html#user-s-guide">User’s guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../readme.html#developer-s-guide">Developer’s guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../readme.html#citations">Citations</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Cookbook</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Run MineSV demo at Biowulf</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#installation">Installation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#start-to-make-the-folder-for-this-tutorial">Start to make the folder for this tutorial</a></li>
<li class="toctree-l3"><a class="reference internal" href="#clone-minesv-from-github">Clone <em>MineSV</em> from github</a></li>
<li class="toctree-l3"><a class="reference internal" href="#add-annotation-data-set-to-minesv">Add annotation data set to MineSV</a></li>
<li class="toctree-l3"><a class="reference internal" href="#download-the-input-sequence-data-set-a-href-https-doi-org-10-5281-zenodo-5217570-img-src-https-zenodo-org-badge-doi-10-5281-zenodo-5217570-svg-alt-doi-a">Download the input sequence data set <span class="raw-html-m2r"><a href="https://doi.org/10.5281/zenodo.5217570"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.5217570.svg" alt="DOI"></a></span></a></li>
<li class="toctree-l3"><a class="reference internal" href="#check-the-current-status">Check the current status</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#configure-the-minesv-run">Configure the <em>MineSV</em> run</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#create-the-directory-working-dir-and-prepare-the-input-file">Create the directory working_dir and prepare the input file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#create-the-file-config-yaml">Create the file config.yaml</a></li>
<li class="toctree-l3"><a class="reference internal" href="#copy-the-scripts-sv-wrapper-sh-and-run-pipeline-sh">Copy the scripts SV_wrapper.sh and run_pipeline.sh</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#run-the-minesv-snakemake-pipeline">Run the MineSV snakemake pipeline</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#start-an-interactive-session-at-biowulf-2-minutes">Start an interactive session at Biowulf (~ 2 minutes)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#generate-the-rulegraph-plot-of-minesv">Generate the rulegraph plot of MineSV</a></li>
<li class="toctree-l3"><a class="reference internal" href="#build-indices-for-the-new-reference-genome-homo-sapiens-chr22-assembly19-fasta-optional-15-minutes">Build indices for the new reference genome Homo_sapiens_chr22_assembly19.fasta (optional, ~15 minutes)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#run-minesv-30-minutes">Run MineSV (&gt;30 minutes)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#inspect-minesv-output-files">Inspect MineSV output files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#the-original-output-files-from-the-sv-callers">The original output files from the SV callers</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sv-comparison-and-annotateion-results-from-minesv">SV comparison and annotateion results from MineSV</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">MineSV</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Run MineSV demo at Biowulf</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/cookbooks/MineSV_demo_at_Biowulf.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="run-minesv-demo-at-biowulf">
<h1>Run MineSV demo at Biowulf<a class="headerlink" href="#run-minesv-demo-at-biowulf" title="Permalink to this headline">¶</a></h1>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This tutorial is to guide users to install and run <em>MineSV</em> using the Slurm workload management at <a class="reference external" href="https://hpc.nih.gov/docs/userguide.html">Biowulf</a>.  We are going to run three SV calls via <em>MineSV</em>: svaba, manta and gridss on one set of normal/tumor paired next-generation sequencing data.</p>
</section>
<section id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<section id="start-to-make-the-folder-for-this-tutorial">
<h3>Start to make the folder for this tutorial<a class="headerlink" href="#start-to-make-the-folder-for-this-tutorial" title="Permalink to this headline">¶</a></h3>
<pre>
export DEMO_DIR=/data/<mark>yourusername</mark>/MineSV_Demo_at_Biowulf
</pre><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir -p <span class="nv">$DEMO_DIR</span><span class="p">;</span> <span class="nb">cd</span> <span class="nv">$_</span>
</pre></div>
</div>
</section>
<section id="clone-minesv-from-github">
<h3>Clone <em>MineSV</em> from github<a class="headerlink" href="#clone-minesv-from-github" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git clone https://github.com/NCI-CGR/MineSV
</pre></div>
</div>
</section>
<section id="add-annotation-data-set-to-minesv">
<h3>Add annotation data set to MineSV<a class="headerlink" href="#add-annotation-data-set-to-minesv" title="Permalink to this headline">¶</a></h3>
<p>Due to the file size limit in the github, we deposit the annotation component of MineSV to Zenodo <span class="raw-html-m2r"><a href="https://doi.org/10.5281/zenodo.5176148"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.5176148.svg" alt="DOI"></a></span>.  And we need to add this component to the MineSV clone before running this snakeamke pipeline.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget -O MineSV_annotation.tgz https://zenodo.org/record/5176148/files/MineSV_annotation.tgz?download<span class="o">=</span><span class="m">1</span>

tar xvfz  MineSV_annotation.tgz -C MineSV

tree MineSV/annotation
MineSV/annotation
├── genes
│   ├── b37
│   │   └── RefSeq.bed
│   ├── hg19
│   │   └── RefSeq.bed
│   └── hg38
│       └── RefSeq.bed
├── genomic_context
│   ├── b37
│   │   ├── RepeatMasker.bed
│   │   ├── SegDups.bed
│   │   └── Telo_Centro.bed
│   ├── hg19
│   │   ├── RepeatMasker.bed
│   │   ├── SegDups.bed
│   │   └── Telo_Centro.bed
│   └── hg38
│       ├── RepeatMasker.bed
│       ├── SegDups.bed
│       └── Telo_Centro.bed
├── human_b37.genome
├── human_hg19.genome
├── human_hg38.genome
└── public_datasets
    ├── b37
    │   ├── 1KG.bed
    │   ├── ClinGen.bed
    │   ├── ClinVar.bed
    │   └── DGV.bed
    ├── hg19
    │   ├── 1KG.bed
    │   ├── ClinGen.bed
    │   ├── ClinVar.bed
    │   └── DGV.bed
    └── hg38
        ├── 1KG.bed
        ├── ClinGen.bed
        ├── ClinVar.bed
        └── DGV.bed

<span class="m">12</span> directories, <span class="m">27</span> files
</pre></div>
</div>
</section>
<section id="download-the-input-sequence-data-set-a-href-https-doi-org-10-5281-zenodo-5217570-img-src-https-zenodo-org-badge-doi-10-5281-zenodo-5217570-svg-alt-doi-a">
<h3>Download the input sequence data set <span class="raw-html-m2r"><a href="https://doi.org/10.5281/zenodo.5217570"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.5217570.svg" alt="DOI"></a></span><a class="headerlink" href="#download-the-input-sequence-data-set-a-href-https-doi-org-10-5281-zenodo-5217570-img-src-https-zenodo-org-badge-doi-10-5281-zenodo-5217570-svg-alt-doi-a" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget -O MineSV_Demo_SeqInput.zip https://zenodo.org/record/5217570/files/MineSV_Demo_SeqInput.zip?download<span class="o">=</span><span class="m">1</span>

unzip MineSV_Demo_SeqInput.zip
</pre></div>
</div>
</section>
<section id="check-the-current-status">
<h3>Check the current status<a class="headerlink" href="#check-the-current-status" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tree -L <span class="m">2</span>
.
├── bam
│   ├── normal.bam
│   ├── normal.bam.bai
│   ├── tumor.bam
│   └── tumor.bam.bai
├── MineSV
│   ├── annotation
│   ├── config.yaml
│   ├── docs
│   ├── LICENSE
│   ├── modules
│   ├── README.md
│   ├── run_pipeline_at_biowulf.sh
│   ├── run_pipeline_at_ccad.sh
│   ├── scripts
│   ├── Snakefile_SV_scaffold
│   ├── SV_wrapper.sh
│   └── TEMPLATE_singularity_recipe
├── MineSV_annotation.tgz
├── MineSV_Demo_SeqInput.zip
└── refGenomes
    └── Homo_sapiens_chr22_assembly19.fasta

<span class="m">7</span> directories, <span class="m">15</span> files
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="configure-the-minesv-run">
<h2>Configure the <em>MineSV</em> run<a class="headerlink" href="#configure-the-minesv-run" title="Permalink to this headline">¶</a></h2>
<section id="create-the-directory-working-dir-and-prepare-the-input-file">
<h3>Create the directory working_dir and prepare the input file<a class="headerlink" href="#create-the-directory-working-dir-and-prepare-the-input-file" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir workding_dir<span class="p">;</span> <span class="nb">cd</span> <span class="nv">$_</span>

cat <span class="s">&lt;&lt;EOF &gt; input.txt</span>
<span class="s">sample1 tumor.bam normal.bam</span>
<span class="s">EOF</span>
</pre></div>
</div>
</section>
<section id="create-the-file-config-yaml">
<h3>Create the file config.yaml<a class="headerlink" href="#create-the-file-config-yaml" title="Permalink to this headline">¶</a></h3>
<p>User may make a copy of the file config.yaml under the folder ../MineSV to the folder worling_dir, and make some necessary changes.  Briefly, we select the analysis mode as ‘TN’ for the somatic SV call from the Tumor-Normal paired samples. And we use three SV callers in this particular demo: svaba, manta and gridss.</p>
<p>The revised config.yaml for this tutorial is as below:</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1">#### Basic analysis parameters ####</span>
<span class="nt">analysisMode</span><span class="p">:</span> <span class="s">&#39;TN&#39;</span>  <span class="c1"># TN, TO, germline, de_novo</span>
<span class="nt">callers</span><span class="p">:</span>            <span class="c1"># available callers: svaba, breakdancer, delly, manta</span>
<span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">svaba</span>
<span class="c1"># - breakdancer</span>
<span class="c1"># - delly</span>
<span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">manta</span>
<span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">gridss</span>

<span class="nt">runMode</span><span class="p">:</span>
  <span class="nt">callAndAnnotate</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">yes</span>
  <span class="nt">callOnly</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">no</span>
  <span class="nt">annotateOnly</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">no</span>
<span class="nt">refGenome</span><span class="p">:</span> <span class="s">&#39;../refGenomes/Homo_sapiens_chr22_assembly19.fasta&#39;</span>
<span class="nt">genomeBuild</span><span class="p">:</span> <span class="s">&#39;b37&#39;</span>  <span class="c1"># for MineSV annotation</span>

<span class="c1">#### Annotation parameters ####</span>
<span class="nt">annotateFile</span><span class="p">:</span> <span class="c1"># only needed if runMode: annotateOnly</span>
<span class="nt">annotationParams</span><span class="p">:</span>
  <span class="nt">interchromPadding</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">50</span>
  <span class="nt">insertionPadding</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">500</span>
  <span class="nt">crossCallerOverlap</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">0.7</span>
  <span class="nt">genomicContextOverlap</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">0.7</span>
  <span class="nt">publicDataOverlap</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">0.7</span>

<span class="c1">#### Samples to analyze ####</span>
<span class="nt">inFile</span><span class="p">:</span> <span class="s">&#39;input.txt&#39;</span>

<span class="c1">#### Directories ####</span>
<span class="nt">inDir</span><span class="p">:</span> <span class="s">&#39;../bam&#39;</span>
<span class="nt">execDir</span><span class="p">:</span> <span class="s">&#39;../MineSV&#39;</span>
<span class="nt">outDir</span><span class="p">:</span> <span class="s">&#39;output&#39;</span>
<span class="nt">logDir</span><span class="p">:</span> <span class="s">&#39;output/logs&#39;</span>
<span class="nt">tempDir</span><span class="p">:</span> <span class="s">&#39;tmp&#39;</span>

<span class="c1">#### Cluster parameters ####</span>
<span class="nt">clusterMode</span><span class="p">:</span> <span class="s">&quot;</span><span class="nv"> </span><span class="s">sbatch</span><span class="nv">  </span><span class="s">-p</span><span class="nv"> </span><span class="s">norm</span><span class="nv"> </span><span class="s">-e</span><span class="nv"> </span><span class="s">logs/slurm-%j.err</span><span class="nv"> </span><span class="s">-o</span><span class="nv"> </span><span class="s">logs/slurm-%j.out</span><span class="nv"> </span><span class="s">--cpus-per-task={threads}</span><span class="nv">  </span><span class="s">-t</span><span class="nv"> </span><span class="s">200:00:00</span><span class="nv"> </span><span class="s">--mem=40g</span><span class="nv"> </span><span class="s">&quot;</span>
<span class="nt">latency</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">600</span>
<span class="nt">maxNumJobs</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">40</span>
<span class="nt">maxThreads</span><span class="p">:</span>
<span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">4</span>
<span class="c1"># - 1</span>
<span class="c1"># - 1</span>
<span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">4</span>
<span class="p p-Indicator">-</span> <span class="l l-Scalar l-Scalar-Plain">8</span>
</pre></div>
</div>
</section>
<section id="copy-the-scripts-sv-wrapper-sh-and-run-pipeline-sh">
<h3>Copy the scripts SV_wrapper.sh and run_pipeline.sh<a class="headerlink" href="#copy-the-scripts-sv-wrapper-sh-and-run-pipeline-sh" title="Permalink to this headline">¶</a></h3>
<p>SV_wrapper.sh is a wrapper shell script to help users to launch the MineSV pipeline, and run_pipeline.sh is another helper script to submit SV_wrapper.sh as a slurm job itself.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$DEMO_DIR</span>/workding_dir

cp ../MineSV/SV_wrapper.sh .
cp ../MineSV/run_pipeline_at_biowulf.sh run_pipeline.sh
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="run-the-minesv-snakemake-pipeline">
<h2>Run the MineSV snakemake pipeline<a class="headerlink" href="#run-the-minesv-snakemake-pipeline" title="Permalink to this headline">¶</a></h2>
<section id="start-an-interactive-session-at-biowulf-2-minutes">
<h3>Start an interactive session at Biowulf (~ 2 minutes)<a class="headerlink" href="#start-an-interactive-session-at-biowulf-2-minutes" title="Permalink to this headline">¶</a></h3>
<p>This process may take several minutes and you will get a new shell prompt something similar to “(base) [<a class="reference external" href="mailto:zhuw10&#37;&#52;&#48;cn0853">zhuw10<span>&#64;</span>cn0853</a> workding_dir]$ “, where cnXXXX indicates the computer node that has just been allocated to you (more details about sinteractive is available at <a class="reference external" href="https://hpc.nih.gov/docs/userguide.html#int">here</a>).  We use the interactive session for the subsequent operations.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sinteractive --time<span class="o">=</span><span class="m">2160</span> --mem<span class="o">=</span>8g --cpus-per-task<span class="o">=</span><span class="m">4</span>
</pre></div>
</div>
</section>
<section id="generate-the-rulegraph-plot-of-minesv">
<h3>Generate the rulegraph plot of MineSV<a class="headerlink" href="#generate-the-rulegraph-plot-of-minesv" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$DEMO_DIR</span>/workding_dir

module load git/2.30.1 python/3.7 perl/5.24.3 snakemake/5.24.1 singularity/3.7.4 samtools/1.11

mkdir -p tmp/dummy output/logs logs

<span class="nv">conf</span><span class="o">=</span>config.yaml snakemake -s ../MineSV/Snakefile_SV_scaffold --rulegraph <span class="p">|</span> dot -Tpng &gt; dag.png

display dag.png <span class="c1"># Try it if you have set up X11 and ssh properly</span>
</pre></div>
</div>
<a class="reference external image-reference" href="../_static/dag.png"><img alt="" src="../_images/dag.png" /></a>
</section>
<section id="build-indices-for-the-new-reference-genome-homo-sapiens-chr22-assembly19-fasta-optional-15-minutes">
<h3>Build indices for the new reference genome Homo_sapiens_chr22_assembly19.fasta (optional, ~15 minutes)<a class="headerlink" href="#build-indices-for-the-new-reference-genome-homo-sapiens-chr22-assembly19-fasta-optional-15-minutes" title="Permalink to this headline">¶</a></h3>
<p>The building of reference genome indices is part of the MineSV pipeline and will be processed by default if any of the indices is not available.  We take this step is to demonstrate that how the running details is wrapped by the scripts <a href="#id1"><span class="problematic" id="id2">**</span></a><em>SV_wrapper.sh</em>** and <a href="#id3"><span class="problematic" id="id4">**</span></a><em>run_pipeline.sh</em><a href="#id5"><span class="problematic" id="id6">**</span></a>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nv">conf</span><span class="o">=</span>config.yaml snakemake --cores <span class="m">10</span>  -p --latency-wait <span class="m">10</span> <span class="se">\</span>
    -s ../MineSV/Snakefile_SV_scaffold  --use-singularity <span class="se">\</span>
    --singularity-args <span class="s2">&quot;-B tmp:/scratch,../refGenomes:/ref,../MineSV/:/exec&quot;</span> <span class="se">\</span>
    --cluster <span class="s2">&quot; sbatch -p norm -e logs/slurm-%j.err -o logs/slurm-%j.out --cpus-per-task={threads} -t 200:00:00 --mem=8g &quot;</span> <span class="se">\</span>
    -U bwa_index_ref fai_index_ref dict_index_ref
</pre></div>
</div>
</section>
<section id="run-minesv-30-minutes">
<h3>Run MineSV (&gt;30 minutes)<a class="headerlink" href="#run-minesv-30-minutes" title="Permalink to this headline">¶</a></h3>
<p>The launch of the MineSV pipeline has been simiplified by the use of the script run_pipeline.sh :</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">cd</span> <span class="nv">$DEMO_DIR</span>/workding_dir
. run_pipeline.sh
</pre></div>
</div>
<p>Users may get the standard output something like the following:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">-] Unloading git 2.30.1  ...</span>
<span class="go">[+] Loading git 2.30.1  ...</span>
<span class="go">[-] Unloading python 3.7  ...</span>
<span class="go">[+] Loading python 3.7  ...</span>
<span class="go">[-] Unloading perl 5.24.3 on cn0853</span>
<span class="go">[+] Loading perl 5.24.3 on cn0853</span>
<span class="go">[-] Unloading snakemake  5.24.1</span>
<span class="go">[+] Loading snakemake  5.24.1</span>
<span class="go">[-] Unloading singularity  3.7.4  on cn0853</span>
<span class="go">[+] Loading singularity  3.7.4  on cn0853</span>
<span class="go">[-] Unloading samtools 1.11  ...</span>
<span class="go">[+] Loading samtools 1.11  ...</span>
<span class="go">Command run: sbatch -t 200:00:00 --export=ALL --mem=32g -p norm -o /data/zhuw10/repeat/MineSV_Demo_at_Biowulf/workding_dir/SV_wrapper.sh.o%j --wrap=&#39;/data/zhuw10/repeat/MineSV_Demo_at_Biowulf/workding_dir/SV_wrapper.sh /data/zhuw10/repeat/MineSV_Demo_at_Biowulf/workding_dir/config.yaml&#39;</span>
<span class="go">21390543</span>
</pre></div>
</div>
</section>
</section>
<hr class="docutils" />
<section id="inspect-minesv-output-files">
<h2>Inspect MineSV output files<a class="headerlink" href="#inspect-minesv-output-files" title="Permalink to this headline">¶</a></h2>
<p>We have already packed the MineSV output from this tutorial run into a zip file, together with the full MineSV pipeline, the sequencing input data and the corresponding reference genome.  The zipped file is available at
.. image:: <a class="reference external" href="https://zenodo.org/badge/DOI/10.5281/zenodo.5213041.svg">https://zenodo.org/badge/DOI/10.5281/zenodo.5213041.svg</a></p>
<blockquote>
<div><dl class="field-list simple">
<dt class="field-odd">target</dt>
<dd class="field-odd"><p><a class="reference external" href="https://doi.org/10.5281/zenodo.5213041">https://doi.org/10.5281/zenodo.5213041</a></p>
</dd>
<dt class="field-even">alt</dt>
<dd class="field-even"><p>DOI</p>
</dd>
</dl>
</div></blockquote>
<p>.</p>
<pre>
export DEMO_FINAL=/data/<mark>yourusername</mark>/MineSV_Demo_Final
</pre><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir -p <span class="nv">$DEMO_FINAL</span><span class="p">;</span> <span class="nb">cd</span> <span class="nv">$_</span>

<span class="c1">### It may take about 2 hours to download the dataset from Zenodo</span>
wget https://zenodo.org/record/5213041/files/MineSV_Demo_at_Biowulf.zip?download<span class="o">=</span><span class="m">1</span> -O MineSV_Demo_at_Biowulf.zip

unzip MineSV_Demo_at_Biowulf.zip
</pre></div>
</div>
<section id="the-original-output-files-from-the-sv-callers">
<h3>The original output files from the SV callers<a class="headerlink" href="#the-original-output-files-from-the-sv-callers" title="Permalink to this headline">¶</a></h3>
<p>The original outputs from the SV callers are available in the folder <a href="#id7"><span class="problematic" id="id8">**</span></a><em>&lt;caller&amp;gt;_TN</em>** under the working directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tree -L <span class="m">2</span> MineSV_Demo_at_Biowulf/working_dir/*_TN/
MineSV_Demo_at_Biowulf/working_dir/gridss_TN/
├── gridss.tmp.sample1.vcf.gz.tbi
├── sample1.assembly.bam
├── sample1_full.vcf.gz
├── sample1_full.vcf.gz.tbi
├── sample1_somatic.vcf.gz
├── sample1_somatic.vcf.gz.tbi
├── sample1.vcf.gz
└── sample1_workingdir
    ├── gridss.full.20210812_093540.cn0882.47808.log
    ├── libsswjni.so
    ├── normal.bam.gridss.working
    ├── sample1.assembly.bam.gridss.working
    ├── sample1.vcf.gz.gridss.working
    └── tumor.bam.gridss.working
MineSV_Demo_at_Biowulf/working_dir/manta_TN/
├── filtered_bams
│   ├── sample1_N.bam
│   ├── sample1_N.bam.bai
│   ├── sample1_T.bam
│   └── sample1_T.bam.bai
└── sample1
    ├── results
    ├── runWorkflow.py
    ├── runWorkflow.py.config.pickle
    ├── workflow.error.log.txt
    ├── workflow.exitcode.txt
    └── workflow.warning.log.txt
MineSV_Demo_at_Biowulf/working_dir/svaba_TN/
└── calls
    ├── sample1.alignments.txt.gz
    ├── sample1.bps.txt.gz
    ├── sample1.contigs.bam
    ├── sample1.discordant.txt.gz
    ├── sample1.log
    ├── sample1.svaba.germline.indel.vcf
    ├── sample1.svaba.germline.sv.vcf
    ├── sample1.svaba.somatic.indel.vcf
    ├── sample1.svaba.somatic.sv.vcf
    ├── sample1.svaba.unfiltered.germline.indel.vcf
    ├── sample1.svaba.unfiltered.germline.sv.vcf
    ├── sample1.svaba.unfiltered.somatic.indel.vcf
    └── sample1.svaba.unfiltered.somatic.sv.vcf

<span class="m">9</span> directories, <span class="m">31</span> files
</pre></div>
</div>
</section>
<section id="sv-comparison-and-annotateion-results-from-minesv">
<h3>SV comparison and annotateion results from MineSV<a class="headerlink" href="#sv-comparison-and-annotateion-results-from-minesv" title="Permalink to this headline">¶</a></h3>
<p>Running in callAndAnnotate or callOnly modes will create the directory output/compare_and_annotate. This directory contains the files intrachromosomal<em>SVs</em>&lt;sample&amp;gt;, which contain the superset of SVs from all callers. NOTE: this is under development, and inter-chromosomal SVs are being integrated currently.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>tree -L <span class="m">1</span> MineSV_Demo_at_Biowulf/working_dir/output/compare_and_annotate/
MineSV_Demo_at_Biowulf/working_dir/output/compare_and_annotate/
├── gridss
├── interchromosomal_SVs_sample1
├── intrachromosomal_SVs_sample1
├── manta
└── svaba

<span class="m">3</span> directories, <span class="m">2</span> files
</pre></div>
</div>
</section>
</section>
</section>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="../readme.html" class="btn btn-neutral float-left" title="Overview" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Bari Ballew, Wei Zhu.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>